In [1]:
# For Data Analysis

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
pd.set_option('display.max_columns', 500)

# For Data Visualization

import plotly as py
import plotly.graph_objs as go
import plotly.express as px
import plotly.figure_factory as ff
import plotly.io as pio
import plotly.offline as pyo

from plotly.subplots import make_subplots
from plotly.offline import plot

# For Additional Use
import datetime
import calendar

# Template
pio.templates.default = "plotly_dark"

# Ignore Warnings
import warnings
warnings.filterwarnings("ignore")

In [2]:
data = pd.read_csv('global-data-on-sustainable-energy.csv')
data.head()

Entity  Year  Access to electricity (% of population)  \
0  Afghanistan  2000                                 1.613591   
1  Afghanistan  2001                                 4.074574   
2  Afghanistan  2002                                 9.409158   
3  Afghanistan  2003                                14.738506   
4  Afghanistan  2004                                20.064968   

   Access to clean fuels for cooking  \
0                                6.2   
1                                7.2   
2                                8.2   
3                                9.5   
4                               10.9   

   Renewable-electricity-generating-capacity-per-capita  \
0                                               9.22      
1                                               8.86      
2                                               8.47      
3                                               8.09      
4                                               7.75      

   Financial flows to developing countries (US $)  \
0                                         20000.0   
1                                        130000.0   
2                                       3950000.0   
3                                      25970000.0   
4                                             NaN   

   Renewable energy share in the total final energy consumption (%)  \
0                                              44.99                  
1                                              45.60                  
2                                              37.83                  
3                                              36.66                  
4                                              44.24                  

   Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                 0.16                             0.0   
1                                 0.09                             0.0   
2                                 0.13                             0.0   
3                                 0.31                             0.0   
4                                 0.33                             0.0   

   Electricity from renewables (TWh)  Low-carbon electricity (% electricity)  \
0                               0.31                               65.957440   
1                               0.50                               84.745766   
2                               0.56                               81.159424   
3                               0.63                               67.021280   
4                               0.56                               62.921350   

   Primary energy consumption per capita (kWh/person)  \
0                                          302.59482    
1                                          236.89185    
2                                          210.86215    
3                                          229.96822    
4                                          204.23125    

   Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
0                                               1.64             
1                                               1.74             
2                                               1.40             
3                                               1.40             
4                                               1.20             

   Value_co2_emissions_kt_by_country  \
0                         760.000000   
1                         730.000000   
2                        1029.999971   
3                        1220.000029   
4                        1029.999971   

   Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
0                                       NaN         NaN             NaN   
1                                       NaN         NaN             NaN   
2                                       NaN         NaN      179.426579   
3                                       NaN    8.832278      190.683

In [3]:
data.shape

(3649, 21)

Observation on  NullData

In [4]:
nulldata = pd.DataFrame(data.isnull().sum().sort_values(ascending=False)[:25])
nulldata = nulldata.rename(columns = {0 : 'Null Values'})
nulldata


Null Values
Renewables (% equivalent primary energy)                   2137
Financial flows to developing countries (US $)             2089
Renewable-electricity-generating-capacity-per-c...          931
Value_co2_emissions_kt_by_country                           428
gdp_growth                                                  317
gdp_per_capita                                              282
Energy intensity level of primary energy (MJ/$2...          207
Renewable energy share in the total final energ...          194
Access to clean fuels for cooking                           169
Electricity from nuclear (TWh)                              126
Low-carbon electricity (% electricity)                       42
Electricity from renewables (TWh)                            21
Electricity from fossil fuels (TWh)                          21
Access to electricity (% of population)                      10
Latitude                                                      1
Land Area(Km2)                                                1
Density\n(P/Km2)                                              1
Longitude                                                     1
Primary energy consumption per capita (kWh/person)            0
Year                                                          0
Entity                                                        0

In [5]:
fig_null_data = go.Figure()
fig_null_data.add_trace(go.Bar(
    x = nulldata.index,
    y = nulldata['Null Values'],
    text = nulldata['Null Values'],
    textposition = 'auto'
))
fig_null_data.update_layout(
    title = 'Null Values Count',
    xaxis_tickfont_size = 7,
    template = 'plotly_dark',
)
fig_null_data.show()

In [6]:
data.columns

Index(['Entity', 'Year', 'Access to electricity (% of population)',
       'Access to clean fuels for cooking',
       'Renewable-electricity-generating-capacity-per-capita',
       'Financial flows to developing countries (US $)',
       'Renewable energy share in the total final energy consumption (%)',
       'Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)',
       'Electricity from renewables (TWh)',
       'Low-carbon electricity (% electricity)',
       'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Value_co2_emissions_kt_by_country',
       'Renewables (% equivalent primary energy)', 'gdp_growth',
       'gdp_per_capita', 'Density\n(P/Km2)', 'Land Area(Km2)', 'Latitude',
       'Longitude'],
      dtype='object')

In [7]:
data[data['Financial flows to developing countries (US $)'].notnull()]

Entity  Year  Access to electricity (% of population)  \
0     Afghanistan  2000                                 1.613591   
1     Afghanistan  2001                                 4.074574   
2     Afghanistan  2002                                 9.409158   
3     Afghanistan  2003                                14.738506   
5     Afghanistan  2005                                25.390894   
...           ...   ...                                      ...   
3644     Zimbabwe  2016                                42.561730   
3645     Zimbabwe  2017                                44.178635   
3646     Zimbabwe  2018                                45.572647   
3647     Zimbabwe  2019                                46.781475   
3648     Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
0                                   6.2   
1                                   7.2   
2                                   8.2   
3                                   9.5   
5                                  12.2   
...                                 ...   
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
5                                                  7.51      
...                                                 ...      
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
5                                          9830000.0   
...                                              ...   
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
5                                                 33.88                  
...                                                 ...                  
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0.31                     

In [8]:
data[data['Entity']=='Philippines']

Entity  Year  Access to electricity (% of population)  \
2619  Philippines  2000                                74.661690   
2620  Philippines  2001                                75.722946   
2621  Philippines  2002                                76.776620   
2622  Philippines  2003                                76.600000   
2623  Philippines  2004                                87.600000   
2624  Philippines  2005                                79.915620   
2625  Philippines  2006                                80.180000   
2626  Philippines  2007                                82.013910   
2627  Philippines  2008                                83.300000   
2628  Philippines  2009                                84.300000   
2629  Philippines  2010                                85.397644   
2630  Philippines  2011                                87.200000   
2631  Philippines  2012                                87.080000   
2632  Philippines  2013                                87.500000   
2633  Philippines  2014                                89.720370   
2634  Philippines  2015                                89.080000   
2635  Philippines  2016                                92.066505   
2636  Philippines  2017                                93.000000   
2637  Philippines  2018                                94.448265   
2638  Philippines  2019                                95.645080   
2639  Philippines  2020                                96.842384   

      Access to clean fuels for cooking  \
2619                              37.90   
2620                              39.15   
2621                              39.60   
2622                              40.10   
2623                              40.90   
2624                              40.90   
2625                              41.40   
2626                              40.90   
2627                              40.80   
2628                              40.90   
2629                              40.70   
2630                              40.60   
2631                              40.70   
2632                              40.90   
2633                              42.00   
2634                              42.20   
2635                              43.30   
2636                              44.10   
2637                              45.40   
2638                              46.50   
2639                              48.00   

      Renewable-electricity-generating-capacity-per-capita  \
2619                                              48.53      
2620                                              48.53      
2621                                              50.08      
2622                                              54.30      
2623                                              53.26      
2624                                              52.65      
2625                                              51.72      
2626                                              50.86      
2627                                              50.06      
2628                                              49.67      
2629                                              51.19      
2630                                              50.41      
2631                                              49.88      
2632                                              49.12      
2633                                              52.54      
2634                                              55.08      
2635                                              60.51      
2636                                              61.08      
2637                                              61.82      
2638                                              62.55      
2639                                              62.39      

      Financial flows to developing countries (US $)  \
2619                                      12130000.0   
2620                                         30000.0   
2621                              

Data Classification

The analysis will focus on Sout East Asia, a region that consist of :
<ol>
<li>Brunei</li>
<li>Cambodia</li>
<li>Indonesia</li>
<li>Laos</li>
<li>Malaysia</li>
<li>Myanmar</li>
<li>Philippines</li>
<li>Singapore</li>
<li>Thailand</li>
<li>Vietnam</li>
</ol>
However, data for several countries is not available, hence being excluded. The following is the list of available countries for the region.
<ol>
<li>Cambodia</li>
<li>Indonesia</li>
<li>Malaysia</li>
<li>Myanmar</li>
<li>Philippines</li>
<li>Singapore</li>
<li>Thailand</li>
</ol>


In [9]:
sea = ['Cambodia','Indonesia','Malaysia','Myanmar','Philippines','Singapore','Thailand']

In [10]:
data_sea = data[data['Entity'].isin(sea)]
data_sea.head()

Entity  Year  Access to electricity (% of population)  \
567  Cambodia  2000                                16.600000   
568  Cambodia  2001                                14.624778   
569  Cambodia  2002                                18.187100   
570  Cambodia  2003                                19.300000   
571  Cambodia  2004                                25.298384   

     Access to clean fuels for cooking  \
567                                3.6   
568                                4.1   
569                                4.7   
570                                5.3   
571                                5.9   

     Renewable-electricity-generating-capacity-per-capita  \
567                                               0.83      
568                                               0.82      
569                                               0.97      
570                                               0.96      
571                                               1.02      

     Financial flows to developing countries (US $)  \
567                                             NaN   
568                                             NaN   
569                                             NaN   
570                                       4630000.0   
571                                     409150000.0   

     Renewable energy share in the total final energy consumption (%)  \
567                                              81.58                  
568                                              80.51                  
569                                              80.97                  
570                                              79.92                  
571                                              80.69                  

     Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
567                                 0.42                             0.0   
568                                 0.48                             0.0   
569                                 0.66                             0.0   
570                                 0.67                             0.0   
571                                 0.73                             0.0   

     Electricity from renewables (TWh)  \
567                               0.05   
568                               0.03   
569                               0.03   
570                               0.04   
571                               0.03   

     Low-carbon electricity (% electricity)  \
567                               10.638298   
568                                5.882353   
569                                4.347826   
570                                5.633803   
571                                3.947368   

     Primary energy consumption per capita (kWh/person)  \
567                                          739.45550    
568                                          769.40970    
569                                          754.14470    
570                                          803.81635    
571                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Value_co2_emissions_kt_by_country  \
567                        1960.000000   
568                        2150.000000   
569                        2210.000038   
570                        2380.000114   
571                        2380.000114   

     Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
567                                       NaN    9.993580      300.613679   
568                                       NaN    8.148386      321.1

In [11]:
data_sea.shape

(147, 21)

In [12]:
sea_null = pd.DataFrame( data_sea.isnull().sum().sort_values(ascending=False)[:25])
sea_null = sea_null.rename(columns = {0 : 'Null Values'})
sea_null


Null Values
Financial flows to developing countries (US $)               45
Electricity from nuclear (TWh)                               42
Renewables (% equivalent primary energy)                     42
Energy intensity level of primary energy (MJ/$2...            7
Renewable energy share in the total final energ...            7
Value_co2_emissions_kt_by_country                             7
Entity                                                        0
Latitude                                                      0
Land Area(Km2)                                                0
Density\n(P/Km2)                                              0
gdp_per_capita                                                0
gdp_growth                                                    0
Low-carbon electricity (% electricity)                        0
Primary energy consumption per capita (kWh/person)            0
Year                                                          0
Electricity from renewables (TWh)                             0
Electricity from fossil fuels (TWh)                           0
Renewable-electricity-generating-capacity-per-c...            0
Access to clean fuels for cooking                             0
Access to electricity (% of population)                       0
Longitude                                                     0

In [13]:
fig_null_data_sea = go.Figure()
fig_null_data_sea.add_trace(go.Bar(
    x = sea_null.index,
    y = sea_null['Null Values'],
    text = sea_null['Null Values'],
    textposition = 'auto'
    
))
fig_null_data_sea.update_layout(
    title = 'Null Values Count',
    xaxis_tickfont_size = 7,
    template = 'plotly_dark',
)

### Data Visualisation

In [14]:
data_sea.head()

Entity  Year  Access to electricity (% of population)  \
567  Cambodia  2000                                16.600000   
568  Cambodia  2001                                14.624778   
569  Cambodia  2002                                18.187100   
570  Cambodia  2003                                19.300000   
571  Cambodia  2004                                25.298384   

     Access to clean fuels for cooking  \
567                                3.6   
568                                4.1   
569                                4.7   
570                                5.3   
571                                5.9   

     Renewable-electricity-generating-capacity-per-capita  \
567                                               0.83      
568                                               0.82      
569                                               0.97      
570                                               0.96      
571                                               1.02      

     Financial flows to developing countries (US $)  \
567                                             NaN   
568                                             NaN   
569                                             NaN   
570                                       4630000.0   
571                                     409150000.0   

     Renewable energy share in the total final energy consumption (%)  \
567                                              81.58                  
568                                              80.51                  
569                                              80.97                  
570                                              79.92                  
571                                              80.69                  

     Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
567                                 0.42                             0.0   
568                                 0.48                             0.0   
569                                 0.66                             0.0   
570                                 0.67                             0.0   
571                                 0.73                             0.0   

     Electricity from renewables (TWh)  \
567                               0.05   
568                               0.03   
569                               0.03   
570                               0.04   
571                               0.03   

     Low-carbon electricity (% electricity)  \
567                               10.638298   
568                                5.882353   
569                                4.347826   
570                                5.633803   
571                                3.947368   

     Primary energy consumption per capita (kWh/person)  \
567                                          739.45550    
568                                          769.40970    
569                                          754.14470    
570                                          803.81635    
571                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Value_co2_emissions_kt_by_country  \
567                        1960.000000   
568                        2150.000000   
569                        2210.000038   
570                        2380.000114   
571                        2380.000114   

     Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
567                                       NaN    9.993580      300.613679   
568                                       NaN    8.148386      321.1

#### Electricity Analysis

In [15]:
electricity_sea = data_sea[['Entity','Year','Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)','Electricity from renewables (TWh)',]]
electricity_sea.head()

Entity  Year  Electricity from fossil fuels (TWh)  \
567  Cambodia  2000                                 0.42   
568  Cambodia  2001                                 0.48   
569  Cambodia  2002                                 0.66   
570  Cambodia  2003                                 0.67   
571  Cambodia  2004                                 0.73   

     Electricity from nuclear (TWh)  Electricity from renewables (TWh)  
567                             0.0                               0.05  
568                             0.0                               0.03  
569                             0.0                               0.03  
570                             0.0                               0.04  
571                             0.0                               0.03

In [16]:
electricity_sea_sum = electricity_sea.groupby('Year').sum()
electricity_sea_sum['Total Electricity'] = electricity_sea_sum.sum(axis=1)
electricity_sea_sum['% Fossil Fuel'] = (electricity_sea_sum['Electricity from fossil fuels (TWh)']/electricity_sea_sum['Total Electricity'])*100
electricity_sea_sum['% Nuclear'] = (electricity_sea_sum['Electricity from nuclear (TWh)']/electricity_sea_sum['Total Electricity'])*100
electricity_sea_sum['% Renewable'] = (electricity_sea_sum['Electricity from renewables (TWh)']/electricity_sea_sum['Total Electricity'])*100
electricity_sea_sum.head()

Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
Year                                                                        
2000                               281.00                             0.0   
2001                               305.34                             0.0   
2002                               326.35                             0.0   
2003                               346.20                             0.0   
2004                               376.47                             0.0   

      Electricity from renewables (TWh)  Total Electricity  % Fossil Fuel  \
Year                                                                        
2000                              55.07             336.07      83.613533   
2001                              55.17             360.51      84.696680   
2002                              54.58             380.93      85.671908   
2003                              53.89             400.09      86.530531   
2004                              56.20             432.67      87.010886   

      % Nuclear  % Renewable  
Year                          
2000        0.0    16.386467  
2001        0.0    15.303320  
2002        0.0    14.328092  
2003        0.0    13.469469  
2004        0.0    12.989114

In [17]:
fig_electricity_sea_volume = go.Figure()
fig_electricity_sea_volume.add_trace(go.Scatter(x = electricity_sea_sum.index, y = electricity_sea_sum['Total Electricity'], name = 'Total Electricity', mode = 'lines+markers'))
fig_electricity_sea_volume.add_trace(go.Scatter(x = electricity_sea_sum.index, y = electricity_sea_sum['Electricity from fossil fuels (TWh)'], name = ' Fossil Fuel', mode = 'lines+markers'))
fig_electricity_sea_volume.add_trace(go.Scatter(x = electricity_sea_sum.index, y = electricity_sea_sum['Electricity from nuclear (TWh)'], name = 'Nuclear', mode = 'lines+markers'))
fig_electricity_sea_volume.add_trace(go.Scatter(x = electricity_sea_sum.index, y = electricity_sea_sum['Electricity from renewables (TWh)'], name = 'Renewable', mode = 'lines+markers'))
fig_electricity_sea_volume.update_layout( title = 'Electricity Consumption in SEA', xaxis_tickfont_size = 7, template = 'plotly_dark')
fig_electricity_sea_volume.update_xaxes(rangeslider_visible = True)
fig_electricity_sea_volume.show()

In [18]:
fig_electricity_sea_portion = go.Figure()
fig_electricity_sea_portion = fig_electricity_sea_portion.add_trace(go.Bar(x = electricity_sea_sum.index, y = electricity_sea_sum['% Fossil Fuel'], name = 'Fossil Fuel'))
fig_electricity_sea_portion = fig_electricity_sea_portion.add_trace(go.Bar(x = electricity_sea_sum.index, y = electricity_sea_sum['% Nuclear'], name = 'Nuclear'))
fig_electricity_sea_portion = fig_electricity_sea_portion.add_trace(go.Bar(x = electricity_sea_sum.index, y = electricity_sea_sum['% Renewable'], name = 'Renewable'))
fig_electricity_sea_portion = fig_electricity_sea_portion.update_layout( title = 'Electricity Consumption Mix in SEA', xaxis_tickfont_size = 7, template = 'plotly_dark', barmode = 'stack')
fig_electricity_sea_portion.show()

In [19]:
electricity_sea_country_sum= electricity_sea[['Year','Entity','Electricity from fossil fuels (TWh)','Electricity from renewables (TWh)']]
electricity_sea_country_sum.fillna(0,inplace=True)
electricity_sea_country_sum['Total Electricity'] = electricity_sea_country_sum['Electricity from fossil fuels (TWh)'] + electricity_sea_country_sum['Electricity from renewables (TWh)']
electricity_sea_country_sum.head()


Year    Entity  Electricity from fossil fuels (TWh)  \
567  2000  Cambodia                                 0.42   
568  2001  Cambodia                                 0.48   
569  2002  Cambodia                                 0.66   
570  2003  Cambodia                                 0.67   
571  2004  Cambodia                                 0.73   

     Electricity from renewables (TWh)  Total Electricity  
567                               0.05               0.47  
568                               0.03               0.51  
569                               0.03               0.69  
570                               0.04               0.71  
571                               0.03               0.76

In [20]:
electricity_sea_country_portion = electricity_sea_country_sum[['Year','Entity','Electricity from renewables (TWh)','Electricity from fossil fuels (TWh)','Total Electricity']].reset_index(drop=True)
electricity_sea_country_portion['portion_renewable'] = (electricity_sea_country_portion['Electricity from renewables (TWh)']/electricity_sea_country_portion['Total Electricity'])*100
electricity_sea_country_portion['portion_fossil'] = (electricity_sea_country_portion['Electricity from fossil fuels (TWh)']/electricity_sea_country_portion['Total Electricity'])*100

electricity_sea_country_portion.head()

Year    Entity  Electricity from renewables (TWh)  \
0  2000  Cambodia                               0.05   
1  2001  Cambodia                               0.03   
2  2002  Cambodia                               0.03   
3  2003  Cambodia                               0.04   
4  2004  Cambodia                               0.03   

   Electricity from fossil fuels (TWh)  Total Electricity  portion_renewable  \
0                                 0.42               0.47          10.638298   
1                                 0.48               0.51           5.882353   
2                                 0.66               0.69           4.347826   
3                                 0.67               0.71           5.633803   
4                                 0.73               0.76           3.947368   

   portion_fossil  
0       89.361702  
1       94.117647  
2       95.652174  
3       94.366197  
4       96.052632

In [70]:
def draw_graph(countries,column,data_frame):
    fig_electricity_sea_country_volume = go.Figure()
    
    for country in countries:
        country_data = data_frame[data_frame['Entity'] == country]
        fig_electricity_sea_country_volume.add_trace(go.Scatter(x = country_data['Year'], y = country_data[column], name = country, mode = 'lines+markers'))
    fig_electricity_sea_country_volume.update_layout( title = 'Electricity Consumption in SEA' + ' : ' + column, xaxis_tickfont_size = 7, template = 'plotly_dark')
    fig_electricity_sea_country_volume.show()

In [71]:
draw_graph(sea,'Total Electricity',electricity_sea_country_sum)

In [72]:
draw_graph(sea,'Electricity from fossil fuels (TWh)',electricity_sea_country_sum)

In [73]:
draw_graph(sea,'Electricity from renewables (TWh)',electricity_sea_country_sum)

In [25]:
electricity_sea_country_growth = electricity_sea_country_sum[['Year','Entity','Total Electricity']]
electricity_sea_country_growth.head()

Year    Entity  Total Electricity
567  2000  Cambodia               0.47
568  2001  Cambodia               0.51
569  2002  Cambodia               0.69
570  2003  Cambodia               0.71
571  2004  Cambodia               0.76

In [26]:
def draw_graph_bar(countries,column1,column2,dataframe):
    for country in countries:
        country_data = dataframe[dataframe['Entity'] == country]
        fig_electricity_sea_country_volume = go.Figure()
        fig_electricity_sea_country_volume.add_trace(go.Bar(x = country_data['Year'], y = country_data[column1], name = 'Renewable Energy'))
        fig_electricity_sea_country_volume.add_trace(go.Bar(x = country_data['Year'], y = country_data[column2], name = 'Fossil Energy'))
        fig_electricity_sea_country_volume.update_layout( title = 'Electricity Production Mix in SEA' + ' : ' + country, xaxis_tickfont_size = 7, template = 'plotly_dark', barmode = 'stack')
        fig_electricity_sea_country_volume.show()

In [27]:
draw_graph_bar(sea,'portion_renewable','portion_fossil',electricity_sea_country_portion)

In [28]:
electricity_sea.head()

Entity  Year  Electricity from fossil fuels (TWh)  \
567  Cambodia  2000                                 0.42   
568  Cambodia  2001                                 0.48   
569  Cambodia  2002                                 0.66   
570  Cambodia  2003                                 0.67   
571  Cambodia  2004                                 0.73   

     Electricity from nuclear (TWh)  Electricity from renewables (TWh)  
567                             0.0                               0.05  
568                             0.0                               0.03  
569                             0.0                               0.03  
570                             0.0                               0.04  
571                             0.0                               0.03

#### Primary Energy Analysis

In [29]:
data_sea.head()

Entity  Year  Access to electricity (% of population)  \
567  Cambodia  2000                                16.600000   
568  Cambodia  2001                                14.624778   
569  Cambodia  2002                                18.187100   
570  Cambodia  2003                                19.300000   
571  Cambodia  2004                                25.298384   

     Access to clean fuels for cooking  \
567                                3.6   
568                                4.1   
569                                4.7   
570                                5.3   
571                                5.9   

     Renewable-electricity-generating-capacity-per-capita  \
567                                               0.83      
568                                               0.82      
569                                               0.97      
570                                               0.96      
571                                               1.02      

     Financial flows to developing countries (US $)  \
567                                             NaN   
568                                             NaN   
569                                             NaN   
570                                       4630000.0   
571                                     409150000.0   

     Renewable energy share in the total final energy consumption (%)  \
567                                              81.58                  
568                                              80.51                  
569                                              80.97                  
570                                              79.92                  
571                                              80.69                  

     Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
567                                 0.42                             0.0   
568                                 0.48                             0.0   
569                                 0.66                             0.0   
570                                 0.67                             0.0   
571                                 0.73                             0.0   

     Electricity from renewables (TWh)  \
567                               0.05   
568                               0.03   
569                               0.03   
570                               0.04   
571                               0.03   

     Low-carbon electricity (% electricity)  \
567                               10.638298   
568                                5.882353   
569                                4.347826   
570                                5.633803   
571                                3.947368   

     Primary energy consumption per capita (kWh/person)  \
567                                          739.45550    
568                                          769.40970    
569                                          754.14470    
570                                          803.81635    
571                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Value_co2_emissions_kt_by_country  \
567                        1960.000000   
568                        2150.000000   
569                        2210.000038   
570                        2380.000114   
571                        2380.000114   

     Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
567                                       NaN    9.993580      300.613679   
568                                       NaN    8.148386      321.1

In [30]:
data_sea.columns

Index(['Entity', 'Year', 'Access to electricity (% of population)',
       'Access to clean fuels for cooking',
       'Renewable-electricity-generating-capacity-per-capita',
       'Financial flows to developing countries (US $)',
       'Renewable energy share in the total final energy consumption (%)',
       'Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)',
       'Electricity from renewables (TWh)',
       'Low-carbon electricity (% electricity)',
       'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Value_co2_emissions_kt_by_country',
       'Renewables (% equivalent primary energy)', 'gdp_growth',
       'gdp_per_capita', 'Density\n(P/Km2)', 'Land Area(Km2)', 'Latitude',
       'Longitude'],
      dtype='object')

In [31]:
primary_energy_sea = data_sea[['Entity','Year','Primary energy consumption per capita (kWh/person)','Energy intensity level of primary energy (MJ/$2017 PPP GDP)','Renewables (% equivalent primary energy)','gdp_per_capita']]
primary_energy_sea.fillna(0,inplace=True)
primary_energy_sea.head()

Entity  Year  Primary energy consumption per capita (kWh/person)  \
567  Cambodia  2000                                          739.45550    
568  Cambodia  2001                                          769.40970    
569  Cambodia  2002                                          754.14470    
570  Cambodia  2003                                          803.81635    
571  Cambodia  2004                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Renewables (% equivalent primary energy)  gdp_per_capita  
567                                       0.0      300.613679  
568                                       0.0      321.150224  
569                                       0.0      338.987477  
570                                       0.0      362.335482  
571                                       0.0      408.513639

In [32]:
primary_energy_sea.isnull().sum()

Entity                                                         0
Year                                                           0
Primary energy consumption per capita (kWh/person)             0
Energy intensity level of primary energy (MJ/$2017 PPP GDP)    0
Renewables (% equivalent primary energy)                       0
gdp_per_capita                                                 0
dtype: int64

In [33]:
#### draw graph fuction for each of the primary energy column

def draw_primary_graph(countries,column,data_frame):
    fig_primary_energy_sea_country_volume = go.Figure()
    
    for country in countries:
        country_data = data_frame[data_frame['Entity'] == country]
        fig_primary_energy_sea_country_volume.add_trace(go.Scatter(x = country_data['Year'], y = country_data[column], name = country, mode = 'lines+markers'))
    fig_primary_energy_sea_country_volume.update_layout( title = 'Primary Energy Consumption in SEA' + ' : ' + column, xaxis_tickfont_size = 7, template = 'plotly_dark')
    fig_primary_energy_sea_country_volume.show()

draw_primary_graph(sea,'Primary energy consumption per capita (kWh/person)',primary_energy_sea)

draw_primary_graph(sea,'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',primary_energy_sea)



In [34]:
def double_y_axes_consumption_intentsity(countries,column_1,column_2,dataframe):
    for country in countries:
        country_data = dataframe[dataframe['Entity'] == country]
        fig = make_subplots(specs=[[{"secondary_y": True}]])
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_1], name=f'{country} - {column_1}', mode='lines+markers'), secondary_y=False)
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_2], name=f'{country} - {column_2}', mode='lines+markers'), secondary_y=True)
        fig.update_layout(title=f'Primary Energy Consumption & Energy Intensity - {country}',xaxis_title='Year',yaxis_title=column_1,template='plotly_dark')
        fig.update_yaxes(title_text=column_1, secondary_y=False)
        fig.update_yaxes(title_text=column_2, secondary_y=True)
        fig.update_xaxes(tickfont=dict(size=10), title_font=dict(size=10))
        fig.update_yaxes(tickfont = dict(size=10), title_font = dict(size=10))
        fig.show()

In [35]:
double_y_axes_consumption_intentsity(sea,'Primary energy consumption per capita (kWh/person)','Energy intensity level of primary energy (MJ/$2017 PPP GDP)',primary_energy_sea)

In [36]:
primary_energy_sea.columns

Index(['Entity', 'Year', 'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Renewables (% equivalent primary energy)', 'gdp_per_capita'],
      dtype='object')

In [37]:
def double_y_axes_consumption_gdp(countries,column_1,column_2,dataframe):
    for country in countries:
        country_data = dataframe[dataframe['Entity'] == country]
        fig = make_subplots(specs=[[{"secondary_y": True}]])
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_1], name=f'{country} - {column_1}', mode='lines+markers'), secondary_y=False)
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_2], name=f'{country} - {column_2}', mode='lines+markers'), secondary_y=True)
        fig.update_layout(title=f'Primary Energy Consumption & GDP - {country}',xaxis_title='Year',yaxis_title=column_1,template='plotly_dark')
        fig.update_yaxes(title_text=column_1, secondary_y=False)
        fig.update_yaxes(title_text=column_2, secondary_y=True)
        fig.update_xaxes(tickfont=dict(size=10), title_font=dict(size=10))
        fig.update_yaxes(tickfont = dict(size=10), title_font = dict(size=10))
        fig.show()

In [38]:
double_y_axes_consumption_gdp(sea,'Primary energy consumption per capita (kWh/person)','gdp_per_capita',primary_energy_sea)

#### CO2 Analysis

In [39]:
data_sea.head()

Entity  Year  Access to electricity (% of population)  \
567  Cambodia  2000                                16.600000   
568  Cambodia  2001                                14.624778   
569  Cambodia  2002                                18.187100   
570  Cambodia  2003                                19.300000   
571  Cambodia  2004                                25.298384   

     Access to clean fuels for cooking  \
567                                3.6   
568                                4.1   
569                                4.7   
570                                5.3   
571                                5.9   

     Renewable-electricity-generating-capacity-per-capita  \
567                                               0.83      
568                                               0.82      
569                                               0.97      
570                                               0.96      
571                                               1.02      

     Financial flows to developing countries (US $)  \
567                                             NaN   
568                                             NaN   
569                                             NaN   
570                                       4630000.0   
571                                     409150000.0   

     Renewable energy share in the total final energy consumption (%)  \
567                                              81.58                  
568                                              80.51                  
569                                              80.97                  
570                                              79.92                  
571                                              80.69                  

     Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
567                                 0.42                             0.0   
568                                 0.48                             0.0   
569                                 0.66                             0.0   
570                                 0.67                             0.0   
571                                 0.73                             0.0   

     Electricity from renewables (TWh)  \
567                               0.05   
568                               0.03   
569                               0.03   
570                               0.04   
571                               0.03   

     Low-carbon electricity (% electricity)  \
567                               10.638298   
568                                5.882353   
569                                4.347826   
570                                5.633803   
571                                3.947368   

     Primary energy consumption per capita (kWh/person)  \
567                                          739.45550    
568                                          769.40970    
569                                          754.14470    
570                                          803.81635    
571                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Value_co2_emissions_kt_by_country  \
567                        1960.000000   
568                        2150.000000   
569                        2210.000038   
570                        2380.000114   
571                        2380.000114   

     Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
567                                       NaN    9.993580      300.613679   
568                                       NaN    8.148386      321.1

In [40]:
data_sea.columns

Index(['Entity', 'Year', 'Access to electricity (% of population)',
       'Access to clean fuels for cooking',
       'Renewable-electricity-generating-capacity-per-capita',
       'Financial flows to developing countries (US $)',
       'Renewable energy share in the total final energy consumption (%)',
       'Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)',
       'Electricity from renewables (TWh)',
       'Low-carbon electricity (% electricity)',
       'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Value_co2_emissions_kt_by_country',
       'Renewables (% equivalent primary energy)', 'gdp_growth',
       'gdp_per_capita', 'Density\n(P/Km2)', 'Land Area(Km2)', 'Latitude',
       'Longitude'],
      dtype='object')

In [41]:
co2_sea = data_sea[['Entity','Year','Value_co2_emissions_kt_by_country','gdp_growth',]]
co2_sea.head()

Entity  Year  Value_co2_emissions_kt_by_country  gdp_growth
567  Cambodia  2000                        1960.000000    9.993580
568  Cambodia  2001                        2150.000000    8.148386
569  Cambodia  2002                        2210.000038    6.578940
570  Cambodia  2003                        2380.000114    8.505896
571  Cambodia  2004                        2380.000114   10.340529

In [42]:
def co2_gdp_sea(countries,column_1,column_2,dataframe):
    for country in countries:
        country_data = dataframe[dataframe['Entity'] == country]
        fig = make_subplots(specs=[[{"secondary_y": True}]])
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_1], name=f'{country} - C02 Emission', mode='lines+markers'), secondary_y=False)
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_2], name=f'{country} - GDP Growth', mode='lines+markers'), secondary_y=True)
        fig.update_layout(title=f'CO2 Emissions - {country}',xaxis_title='Year',yaxis_title=column_1,template='plotly_dark')
        fig.update_yaxes(title_text=column_1, secondary_y=False)
        fig.update_yaxes(title_text=column_2, secondary_y=True)
        fig.update_xaxes(tickfont=dict(size=10), title_font=dict(size=10))
        fig.update_yaxes(tickfont = dict(size=10), title_font = dict(size=10))
        fig.show()

co2_gdp_sea(sea,'Value_co2_emissions_kt_by_country','gdp_growth',co2_sea)

### Analysis

#### Electricity Consumption in South East Asia

Electricity Consumption in Southeast Asia 

Total electricity consumption throughout the region shows an increasing trend throughout 2000 to 2019, before collapsing in the year 2020 due to Covid-19 outbreak. In 2000, the region produced 336.07 TWh and by 2005, it increased to 452 TWh, an increase of 34.5% throughout the region. By the of the decade, its consumption increased by 74.37% from early decade to stand at 568.61 TWh. 

Between 2000 to 2005, Indonesia is the largest electricity consumer in the region. Its electricity consumption increased from 98.03 TWh to 132.88 TWh, an increase of 35.56% within five years. However, Thailand the second largest electricity consumer in the region grew at faster rate, increased from 89.53 TWh to 123 TWh, an increase of 37% within 5 years. Total electricity consumption in Malaysia grew at even faster rate. It grew at 39.62% within 5 years, from 70.47 TWh to 98.39 TWh. 

In this early decade, electricity consumption in the region is driven by fossil fuel consumption. Fossil fuel formed more than 80% of the electricity consumed in the region. In 2000, 281 TWh or 83.61% of the electricity consumed was generated by fossil fuel. The number keeps increasing. By 2005, it increased by 40.17% to 393.89 TWh. The volume is 87.14% of total electricity consumption in the region. By the end of the decade, 494.27 TWh of fossil fuel was consumed. However, its part of total electricity consumption slightly decreased to 86.93% due to the increase in renewable energy consumption. 

Renewable energy in the region plays crucial roles in the region. In 2000, 16.39% or 55.07 TWh of renewable energy being consumed.  It grew slightly by 5.56% to 58.13 by 2005 TWh. However, renewable energy shares in total electricity consumption mix fell slightly to 12.86 TWh. In 2010, its consumption is at 74.34 TWh, an increase of 35% from the early decade. The volume makes up 13.07% of total electricity consumption in the region. 

In the second decade, total electricity consumption in the region keeps increasing. By 2015, it increased by 24.93% to 710.36 TWh and grow further to reach its peak at 854.5 TWh in 2019. Covid-19 outbreak in 2020 however brought down total electricity consumption by 3.37% to 825.74 TWh in 2020. 

 

In the decade between 2010 to 2020, renewable energy grew at a faster rate than fossil fuels. Between 2010 and 2015, renewable energy consumption increased by 28.58% to 95.59 TWh while fossil fuels grew at slightly lower rates, 24.37% to 614.77 TWh. The period between 2015 to 2020 saw renewable energy consumption grew rapidly at 54.90% to reach its peak at 148.15 TWh in 2020. The largest increase was recorded between 2015 to 2018 when its consumption grew by 51.20% from 95.59 TWh in 2020 to 144.54 TWh in 2018. Fossil fuels consumption meanwhile grew at a much slower rate as it increased by 15.6% to its peak at 710.66 TWh in 2019. Fossil fuels consumption declined in 2020 due to pandemic outbreak to 677.59 TWh. In contrast, renewable energy consumption was not affected by the pandemic outbreak as its consumption reached its peak in 2020. 

Faster renewable energy consumption growth led to the increase of its share in total electricity consumption mix, particularly between 2015 to 2020. In 2010, it contributed to 13.07% of total electricity consumption. By 2015, its shares increased slightly to 13.46% in 2015. Rapid increase in renewable energy consumption led to an increase in its share in total electricity consumption. By 2020, it represented 17.94% of regional electricity consumption. 

Indonesia, Malaysia, and Thailand remained three largest electricity consumers in the region. All three countries showed a similar uptrend throughout the second decade of 2010 to 2020, except for 2020. However, total electricity consumption in Indonesia increased at faster rates compared with Malaysia and Thailand. Between 2010 to 2015, total electricity consumption in Indonesia increased by 37.05% to 243.27 TWh. In comparison, electricity consumption in Malaysia increased by 10.93% and Thailand by 20.3% between the same period. By 2019, total electricity consumption in Indonesia increased by 21.07% to its peak at 295.43 TWh. Meanwhile, total electricity consumption in Malaysia and Thailand grew by 14.32% and 18.89% respectively. 

Rapid increases in the total electricity consumption in Indonesia was fulfilled by the increase in consumption of fossil fuels and renewable energy. Between 2010 to 2015, fossil fuels were the main driver for total electricity consumption. Fossil fuels consumption within this period grew by 46.77%, from 142.88 TWh to 209.71 TWh. Within this period, renewable energy consumption declined by 3.09%, from 34.63 TWh to 33.56 TWh. Renewable energy consumption in Indonesia increased rapidly between 2015 to 2018, increased by 44.16%, from 33.56 TWh to 48.38 TWh and further increased to 52.91 TWh in 2020, defying the decline in the overall electricity consumption due to the outbreak of Covid-19. Between 2015 to 2019, fossil fuel consumption in Indonesia increased by 17.97% to reach its peak, 247.39 TWh in 2019. 

 

### Trial Start

In [43]:
electricity_sea_country = electricity_sea
#electricity_sea_country = electricity_sea_country.set_index('Year')
electricity_sea_country['Total Electricity Country Sum'] = electricity_sea_country[['Electricity from fossil fuels (TWh)','Electricity from nuclear (TWh)','Electricity from renewables (TWh)']].sum(axis=1)
electricity_sea_country.tail()


Entity  Year  Electricity from fossil fuels (TWh)  \
3266  Thailand  2016                               161.79   
3267  Thailand  2017                               161.88   
3268  Thailand  2018                               156.26   
3269  Thailand  2019                               162.59   
3270  Thailand  2020                               154.52   

      Electricity from nuclear (TWh)  Electricity from renewables (TWh)  \
3266                             0.0                              15.97   
3267                             0.0                              19.92   
3268                             0.0                              25.84   
3269                             0.0                              28.02   
3270                             0.0                              24.73   

      Total Electricity Country Sum  
3266                         177.76  
3267                         181.80  
3268                         182.10  
3269                         190.61  
3270                         179.25

In [44]:
electricity_sea_country.columns

Index(['Entity', 'Year', 'Electricity from fossil fuels (TWh)',
       'Electricity from nuclear (TWh)', 'Electricity from renewables (TWh)',
       'Total Electricity Country Sum'],
      dtype='object')

In [45]:
trial0 = electricity_sea_country[['Year','Entity','Total Electricity Country Sum']]
trial0 = trial0.set_index('Year')
trial0.tail()

Entity  Total Electricity Country Sum
Year                                         
2016  Thailand                         177.76
2017  Thailand                         181.80
2018  Thailand                         182.10
2019  Thailand                         190.61
2020  Thailand                         179.25

In [46]:
fig_trial0 = go.Figure()
fig_trial0 = fig_trial0.add_trace(go.Scatter(x = trial0.index, y = trial0[trial0['Entity']=='Philippines']['Total Electricity Country Sum'], name = 'Total Electricity', mode = 'lines+markers'))
fig_trial0.update_layout( title = 'Electricity Production in Philippines', xaxis_tickfont_size = 7, template = 'plotly_dark')
fig_trial0.show()

In [47]:
def draw_graph(countries, data_frame):
    for country in countries:
        country_data = data_frame[data_frame['Entity'] == country]
        fig_trial0 = go.Figure()
        fig_trial0.add_trace(go.Scatter(x=country_data.index, y=country_data['Total Electricity Country Sum'], name=country, mode='lines+markers'))
        fig_trial0.update_layout(title='Electricity Production in '  +  country, xaxis_title='Year', yaxis_title='Total Electricity', xaxis_tickfont_size=7, template='plotly_dark')
        fig_trial0.show()

In [48]:
def draw_graph(countries, data):
    fig_trial0 = go.Figure()
    
    for country in countries:
        country_data = data[data['Entity'] == country]
        fig_trial0.add_trace(go.Scatter(x=country_data.index, y=country_data['Total Electricity Country Sum'], name=country, mode='lines+markers'))
        
    fig_trial0.update_layout(title='Electricity Production in SEA', xaxis_title='Year', yaxis_title='Total Electricity', xaxis_tickfont_size=7, template='plotly_dark')
    fig_trial0.show()

In [49]:
draw_graph(sea, trial0)

In [50]:
primary_energy_sea.head()

Entity  Year  Primary energy consumption per capita (kWh/person)  \
567  Cambodia  2000                                          739.45550    
568  Cambodia  2001                                          769.40970    
569  Cambodia  2002                                          754.14470    
570  Cambodia  2003                                          803.81635    
571  Cambodia  2004                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Renewables (% equivalent primary energy)  gdp_per_capita  
567                                       0.0      300.613679  
568                                       0.0      321.150224  
569                                       0.0      338.987477  
570                                       0.0      362.335482  
571                                       0.0      408.513639

In [51]:
primary_energy_sea[primary_energy_sea['Entity']=='Malaysia']

Entity  Year  Primary energy consumption per capita (kWh/person)  \
2017  Malaysia  2000                                          26639.928    
2018  Malaysia  2001                                          26582.326    
2019  Malaysia  2002                                          27303.566    
2020  Malaysia  2003                                          28867.148    
2021  Malaysia  2004                                          30459.574    
2022  Malaysia  2005                                          32213.139    
2023  Malaysia  2006                                          31831.441    
2024  Malaysia  2007                                          33221.490    
2025  Malaysia  2008                                          33712.133    
2026  Malaysia  2009                                          32003.473    
2027  Malaysia  2010                                          32487.746    
2028  Malaysia  2011                                          32524.594    
2029  Malaysia  2012                                          35010.120    
2030  Malaysia  2013                                          36062.543    
2031  Malaysia  2014                                          35862.223    
2032  Malaysia  2015                                          35878.008    
2033  Malaysia  2016                                          37199.580    
2034  Malaysia  2017                                          37162.887    
2035  Malaysia  2018                                          37333.030    
2036  Malaysia  2019                                          37823.500    
2037  Malaysia  2020                                          35997.363    

      Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
2017                                               5.48             
2018                                               5.67             
2019                                               5.56             
2020                                               5.63             
2021                                               5.68             
2022                                               5.85             
2023                                               5.57             
2024                                               5.59             
2025                                               5.65             
2026                                               5.42             
2027                                               5.24             
2028                                               5.19             
2029                                               4.99             
2030                                               5.38             
2031                                               5.18             
2032                                               4.72             
2033                                               4.69             
2034                                               4.28             
2035                                               4.50             
2036                                               4.25             
2037                                               0.00             

      Renewables (% equivalent primary energy)  gdp_per_capita  
2017                                  3.681303     4043.662923  
2018                                  3.139982     3913.429386  
2019                                  2.593757     4165.726104  
2020                                  2.253213     4461.847325  
2021                                  2.319720     4952.213958  
2022                                  2.286207     5587.024780  
2023                                  2.365507     6209.126167  
2024                                  2.138460     7243.456801  
2025                                  2.624564     8474.587762  
2026                                  2.622681     7292.494973  
2027                                  2.325419     9040.568495  
2028                                

In [52]:
def trial2(countries,column_1,column_2,data_frame):
    fig_primary_energy_sea_country_volume = make_subplots(specs=[[{"secondary_y": True}]])
    
    for country in countries:
        country_data = data_frame[data_frame['Entity'] == country]
        fig_primary_energy_sea_country_volume.add_trace(go.Scatter(x = country_data['Year'], y = country_data[column_1], name = country, mode = 'lines+markers'), secondary_y=False)
        fig_primary_energy_sea_country_volume.add_trace(go.Scatter(x = country_data['Year'], y = country_data[column_2], name = country, mode = 'lines+markers'), secondary_y=True)
        fig_primary_energy_sea_country_volume.update_layout( title = 'Primary Energy Consumption i& GDP per Capita in SEA' + ' : ' + country, xaxis_tickfont_size = 7, template = 'plotly_dark')
        fig_primary_energy_sea_country_volume.update_xaxes(title_text=column_1, secondary_y=False)
        fig_primary_energy_sea_country_volume.update_yaxes(title_text=column_2, secondary_y=True)
        fig_primary_energy_sea_country_volume.show()



In [53]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

def trial2(countries, column_1, column_2, data_frame):
    fig = make_subplots(specs=[[{"secondary_y": True}]])
    
    for country in countries:
        country_data = data_frame[data_frame['Entity'] == country]
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_1], name=f'{country} - {column_1}', mode='lines+markers'), secondary_y=False)
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_2], name=f'{country} - {column_2}', mode='lines+markers'), secondary_y=True)
    
    fig.update_layout(title='Primary Energy Consumption and GDP per Capita in SEA', xaxis_title='Year',yaxis_title=column_1,template='plotly_dark')
    
    fig.update_yaxes(title_text=column_1, secondary_y=False)
    fig.update_yaxes(title_text=column_2, secondary_y=True)
    
    fig.show()





In [54]:
import plotly.graph_objects as go

def trial2(countries, column_1, column_2, data_frame):
    for country in countries:
        country_data = data_frame[data_frame['Entity'] == country]
        fig = go.Figure()
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_1], name=f'{country} - {column_1}', mode='lines+markers',))
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_2], name=f'{country} - {column_2}', mode='lines+markers'))
        fig.update_layout(title=f'Primary Energy Consumption and GDP per Capita in SEA - {country}',xaxis_title='Year',yaxis_title=column_1,template='plotly_dark')
        fig.show()




In [55]:
def trial3(countries,column_1,column_2,dataframe):
    for country in countries:
        country_data = dataframe[dataframe['Entity'] == country]
        fig = make_subplots(specs=[[{"secondary_y": True}]])
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_1], name=f'{country} - {column_1}', mode='lines+markers'), secondary_y=False)
        fig.add_trace(go.Scatter(x=country_data['Year'], y=country_data[column_2], name=f'{country} - {column_2}', mode='lines+markers'), secondary_y=True)
        fig.update_layout(title=f'Primary Energy Consumption and GDP per Capita in SEA - {country}',xaxis_title='Year',yaxis_title=column_1,template='plotly_dark')
        fig.update_yaxes(title_text=column_1, secondary_y=False)
        fig.update_yaxes(title_text=column_2, secondary_y=True)
        fig.show()

In [56]:
trial3(sea,'Primary energy consumption per capita (kWh/person)','Energy intensity level of primary energy (MJ/$2017 PPP GDP)',primary_energy_sea)

In [57]:
data_sea.head()

Entity  Year  Access to electricity (% of population)  \
567  Cambodia  2000                                16.600000   
568  Cambodia  2001                                14.624778   
569  Cambodia  2002                                18.187100   
570  Cambodia  2003                                19.300000   
571  Cambodia  2004                                25.298384   

     Access to clean fuels for cooking  \
567                                3.6   
568                                4.1   
569                                4.7   
570                                5.3   
571                                5.9   

     Renewable-electricity-generating-capacity-per-capita  \
567                                               0.83      
568                                               0.82      
569                                               0.97      
570                                               0.96      
571                                               1.02      

     Financial flows to developing countries (US $)  \
567                                             NaN   
568                                             NaN   
569                                             NaN   
570                                       4630000.0   
571                                     409150000.0   

     Renewable energy share in the total final energy consumption (%)  \
567                                              81.58                  
568                                              80.51                  
569                                              80.97                  
570                                              79.92                  
571                                              80.69                  

     Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
567                                 0.42                             0.0   
568                                 0.48                             0.0   
569                                 0.66                             0.0   
570                                 0.67                             0.0   
571                                 0.73                             0.0   

     Electricity from renewables (TWh)  \
567                               0.05   
568                               0.03   
569                               0.03   
570                               0.04   
571                               0.03   

     Low-carbon electricity (% electricity)  \
567                               10.638298   
568                                5.882353   
569                                4.347826   
570                                5.633803   
571                                3.947368   

     Primary energy consumption per capita (kWh/person)  \
567                                          739.45550    
568                                          769.40970    
569                                          754.14470    
570                                          803.81635    
571                                          813.56445    

     Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
567                                               7.91             
568                                               7.51             
569                                               7.05             
570                                               6.69             
571                                               6.23             

     Value_co2_emissions_kt_by_country  \
567                        1960.000000   
568                        2150.000000   
569                        2210.000038   
570                        2380.000114   
571                        2380.000114   

     Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
567                                       NaN    9.993580      300.613679   
568                                       NaN    8.148386      321.1

In [58]:
electricity_sea_sum.head()

Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
Year                                                                        
2000                               281.00                             0.0   
2001                               305.34                             0.0   
2002                               326.35                             0.0   
2003                               346.20                             0.0   
2004                               376.47                             0.0   

      Electricity from renewables (TWh)  Total Electricity  % Fossil Fuel  \
Year                                                                        
2000                              55.07             336.07      83.613533   
2001                              55.17             360.51      84.696680   
2002                              54.58             380.93      85.671908   
2003                              53.89             400.09      86.530531   
2004                              56.20             432.67      87.010886   

      % Nuclear  % Renewable  
Year                          
2000        0.0    16.386467  
2001        0.0    15.303320  
2002        0.0    14.328092  
2003        0.0    13.469469  
2004        0.0    12.989114

In [59]:
electricity_sea.head()

Entity  Year  Electricity from fossil fuels (TWh)  \
567  Cambodia  2000                                 0.42   
568  Cambodia  2001                                 0.48   
569  Cambodia  2002                                 0.66   
570  Cambodia  2003                                 0.67   
571  Cambodia  2004                                 0.73   

     Electricity from nuclear (TWh)  Electricity from renewables (TWh)  \
567                             0.0                               0.05   
568                             0.0                               0.03   
569                             0.0                               0.03   
570                             0.0                               0.04   
571                             0.0                               0.03   

     Total Electricity Country Sum  
567                           0.47  
568                           0.51  
569                           0.69  
570                           0.71  
571                           0.76

In [60]:
trial4 = pd.DataFrame(electricity_sea_sum['Total Electricity'])
trial4 = trial4.reset_index()
trial4.tail()

Year  Total Electricity
16  2016             754.87
17  2017             779.56
18  2018             817.14
19  2019             854.50
20  2020             825.74

In [61]:
trial4['Total Electricity']

0     336.07
1     360.51
2     380.93
3     400.09
4     432.67
5     452.02
6     470.07
7     493.29
8     512.29
9     523.77
10    568.61
11    589.48
12    629.52
13    661.48
14    689.11
15    710.36
16    754.87
17    779.56
18    817.14
19    854.50
20    825.74
Name: Total Electricity, dtype: float64

In [62]:
electricity_sea.columns

Index(['Entity', 'Year', 'Electricity from fossil fuels (TWh)',
       'Electricity from nuclear (TWh)', 'Electricity from renewables (TWh)',
       'Total Electricity Country Sum'],
      dtype='object')

In [63]:
trial4_cambodia = electricity_sea[electricity_sea['Entity'] == 'Cambodia'].drop('Electricity from nuclear (TWh)', axis=1).reset_index()
trial4_cambodia


index    Entity  Year  Electricity from fossil fuels (TWh)  \
0     567  Cambodia  2000                                 0.42   
1     568  Cambodia  2001                                 0.48   
2     569  Cambodia  2002                                 0.66   
3     570  Cambodia  2003                                 0.67   
4     571  Cambodia  2004                                 0.73   
5     572  Cambodia  2005                                 0.86   
6     573  Cambodia  2006                                 0.98   
7     574  Cambodia  2007                                 1.25   
8     575  Cambodia  2008                                 1.23   
9     576  Cambodia  2009                                 1.02   
10    577  Cambodia  2010                                 0.81   
11    578  Cambodia  2011                                 0.81   
12    579  Cambodia  2012                                 0.51   
13    580  Cambodia  2013                                 0.47   
14    581  Cambodia  2014                                 1.03   
15    582  Cambodia  2015                                 2.10   
16    583  Cambodia  2016                                 2.62   
17    584  Cambodia  2017                                 3.70   
18    585  Cambodia  2018                                 2.96   
19    586  Cambodia  2019                                 3.94   
20    587  Cambodia  2020                                 3.83   

    Electricity from renewables (TWh)  Total Electricity Country Sum  
0                                0.05                           0.47  
1                                0.03                           0.51  
2                                0.03                           0.69  
3                                0.04                           0.71  
4                                0.03                           0.76  
5                                0.04                           0.90  
6                                0.13                           1.11  
7                                0.16                           1.41  
8                                0.16                           1.39  
9                                0.15                           1.17  
10                               0.12                           0.93  
11                               0.18                           0.99  
12                               0.86                           1.37  
13                               1.26                           1.73  
14                               1.96                           2.99  
15                               2.16                           4.26  
16                               2.81                           5.43  
17                               3.07                           6.77  
18                               5.00                           7.96  
19                               4.51                           8.45  
20                               4.62                           8.45

In [64]:
trial4_cambodia = trial4_cambodia.join(trial4[['Year', 'Total Electricity']].set_index('Year'), on='Year')


In [65]:
trial4_cambodia['portion_sum'] = (trial4_cambodia['Total Electricity Country Sum'] / trial4_cambodia['Total Electricity']) * 100
trial4_cambodia

index    Entity  Year  Electricity from fossil fuels (TWh)  \
0     567  Cambodia  2000                                 0.42   
1     568  Cambodia  2001                                 0.48   
2     569  Cambodia  2002                                 0.66   
3     570  Cambodia  2003                                 0.67   
4     571  Cambodia  2004                                 0.73   
5     572  Cambodia  2005                                 0.86   
6     573  Cambodia  2006                                 0.98   
7     574  Cambodia  2007                                 1.25   
8     575  Cambodia  2008                                 1.23   
9     576  Cambodia  2009                                 1.02   
10    577  Cambodia  2010                                 0.81   
11    578  Cambodia  2011                                 0.81   
12    579  Cambodia  2012                                 0.51   
13    580  Cambodia  2013                                 0.47   
14    581  Cambodia  2014                                 1.03   
15    582  Cambodia  2015                                 2.10   
16    583  Cambodia  2016                                 2.62   
17    584  Cambodia  2017                                 3.70   
18    585  Cambodia  2018                                 2.96   
19    586  Cambodia  2019                                 3.94   
20    587  Cambodia  2020                                 3.83   

    Electricity from renewables (TWh)  Total Electricity Country Sum  \
0                                0.05                           0.47   
1                                0.03                           0.51   
2                                0.03                           0.69   
3                                0.04                           0.71   
4                                0.03                           0.76   
5                                0.04                           0.90   
6                                0.13                           1.11   
7                                0.16                           1.41   
8                                0.16                           1.39   
9                                0.15                           1.17   
10                               0.12                           0.93   
11                               0.18                           0.99   
12                               0.86                           1.37   
13                               1.26                           1.73   
14                               1.96                           2.99   
15                               2.16                           4.26   
16                               2.81                           5.43   
17                               3.07                           6.77   
18                               5.00                           7.96   
19                               4.51                           8.45   
20                               4.62                           8.45   

    Total Electricity  portion_sum  
0              336.07     0.139852  
1              360.51     0.141466  
2              380.93     0.181136  
3              400.09     0.177460  
4              432.67     0.175654  
5              452.02     0.199106  
6              470.07     0.236135  
7              493.29     0.285836  
8              512.29     0.271331  
9              523.77     0.223380  
10             568.61     0.163557  
11             589.48     0.167945  
12             629.52     0.217626  
13             661.48     0.261535  
14             689.11     0.433893  
15             710.36     0.599696  
16             754.87     0.719329  
17             779.56     0.868439  
18             817.14     0.974129  
19             854.50     0.988882  
20             825.74     1.023325

In [66]:
trial5 = electricity_sea.drop('Electricity from nuclear (TWh)', axis=1).reset_index()
trial5 = trial5.join(trial4[['Year', 'Total Electricity']].set_index('Year'), on='Year')
trial5['portion_sea'] = (trial5['Total Electricity Country Sum'] / trial5['Total Electricity']) * 100
trial5

index    Entity  Year  Electricity from fossil fuels (TWh)  \
0      567  Cambodia  2000                                 0.42   
1      568  Cambodia  2001                                 0.48   
2      569  Cambodia  2002                                 0.66   
3      570  Cambodia  2003                                 0.67   
4      571  Cambodia  2004                                 0.73   
..     ...       ...   ...                                  ...   
142   3266  Thailand  2016                               161.79   
143   3267  Thailand  2017                               161.88   
144   3268  Thailand  2018                               156.26   
145   3269  Thailand  2019                               162.59   
146   3270  Thailand  2020                               154.52   

     Electricity from renewables (TWh)  Total Electricity Country Sum  \
0                                 0.05                           0.47   
1                                 0.03                           0.51   
2                                 0.03                           0.69   
3                                 0.04                           0.71   
4                                 0.03                           0.76   
..                                 ...                            ...   
142                              15.97                         177.76   
143                              19.92                         181.80   
144                              25.84                         182.10   
145                              28.02                         190.61   
146                              24.73                         179.25   

     Total Electricity  portion_sea  
0               336.07     0.139852  
1               360.51     0.141466  
2               380.93     0.181136  
3               400.09     0.177460  
4               432.67     0.175654  
..                 ...          ...  
142             754.87    23.548426  
143             779.56    23.320848  
144             817.14    22.285043  
145             854.50    22.306612  
146             825.74    21.707801  

[147 rows x 8 columns]

In [67]:
trial5[trial5['Entity'] == 'Thailand']

index    Entity  Year  Electricity from fossil fuels (TWh)  \
126   3250  Thailand  2000                                83.15   
127   3251  Thailand  2001                                88.97   
128   3252  Thailand  2002                                93.51   
129   3253  Thailand  2003                               100.61   
130   3254  Thailand  2004                               109.46   
131   3255  Thailand  2005                               115.58   
132   3256  Thailand  2006                               119.41   
133   3257  Thailand  2007                               122.12   
134   3258  Thailand  2008                               127.43   
135   3259  Thailand  2009                               128.09   
136   3260  Thailand  2010                               141.72   
137   3261  Thailand  2011                               135.31   
138   3262  Thailand  2012                               143.73   
139   3263  Thailand  2013                               148.29   
140   3264  Thailand  2014                               149.26   
141   3265  Thailand  2015                               153.40   
142   3266  Thailand  2016                               161.79   
143   3267  Thailand  2017                               161.88   
144   3268  Thailand  2018                               156.26   
145   3269  Thailand  2019                               162.59   
146   3270  Thailand  2020                               154.52   

     Electricity from renewables (TWh)  Total Electricity Country Sum  \
126                               6.38                          89.53   
127                               6.76                          95.73   
128                               8.07                         101.58   
129                               8.36                         108.97   
130                               7.63                         117.09   
131                               7.42                         123.00   
132                               9.82                         129.23   
133                              10.20                         132.32   
134                               8.95                         136.38   
135                               9.09                         137.18   
136                               8.58                         150.30   
137                              11.83                         147.14   
138                              13.42                         157.15   
139                              12.33                         160.62   
140                              13.68                         162.94   
141                              13.33                         166.73   
142                              15.97                         177.76   
143                              19.92                         181.80   
144                              25.84                         182.10   
145                              28.02                         190.61   
146                              24.73                         179.25   

     Total Electricity  portion_sea  
126             336.07    26.640283  
127             360.51    26.554048  
128             380.93    26.666317  
129             400.09    27.236372  
130             432.67    27.062195  
131             452.02    27.211185  
132             470.07    27.491650  
133             493.29    26.823978  
134             512.29    26.621640  
135             523.77    26.190885  
136             568.61    26.432880  
137             589.48    24.960983  
138             629.52    24.963464  
139             661.48    24.281913  
140             689.11    23.644991  
141             710.36    23.471198  
142             754.87    23.548426  
143             779.56    23.320848  
144             817.14    22.285043  
145             854.50    22.306612  
146             825.74    21.707801

In [68]:
years = trial5['Year'].unique()
years

array([2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
       2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020],
      dtype=int64)

In [69]:
pie_sales_revenue_portion = px.pie(sales_revenue_portion, values='Revenue generated', names='Product type', title='Sales revenue by product type',hole=0.5,\
    color_discrete_sequence=px.colors.qualitative.Pastel)
pie_sales_revenue_portion.update_traces(textposition='inside', textinfo='percent+label')
pie_sales_revenue_portion.show()

NameError: name 'sales_revenue_portion' is not defined

In [ ]:
country_list = data_sea['Entity'].unique()
country_list

array(['Cambodia', 'Indonesia', 'Malaysia', 'Myanmar', 'Philippines',
       'Singapore', 'Thailand'], dtype=object)

In [ ]:
country_list = data_sea['Entity'].unique()
growth_rates = []

for country in country_list:
    country_data = electricity_sea_country_sum[electricity_sea_country_sum['Entity'] == country]
    
    growth_rate = []
    for i in range(len(country_data) - 5):
        growth_rate.append(((country_data.iloc[i + 5]['Total Electricity'] / country_data.iloc[i]['Total Electricity']) ** (1/5) - 1) * 100)
    
    growth_rates.append(growth_rate)

growth_df = pd.DataFrame(growth_rates).transpose()
growth_df.columns = country_list
growth_df.index = range(2005, 2021)

growth_df



Cambodia  Indonesia  Malaysia    Myanmar  Philippines  Singapore  \
2005  13.875142   6.272311  6.902863   3.459005     4.416008   4.902134   
2006  16.828973   5.385104  6.222423   5.780776     3.807594   3.480772   
2007  15.364982   5.594715  6.077024   4.396115     4.184411   3.404638   
2008  14.380316   5.836978  6.543476   3.630058     2.848610   3.315555   
2009   9.012037   5.470301  3.966033   4.788028     2.043379   2.620539   
2010   0.657952   5.962609  4.866990   4.750141     3.608891   3.164602   
2011  -2.262226   6.816461  4.232522  10.117466     4.866814   2.804322   
2012  -0.573926   7.083706  4.360685  11.469113     5.004924   2.349165   
2013   4.473528   7.630311  4.397775  13.609833     5.230383   2.544142   
2014  20.641594   7.841971  4.928648  15.470757     5.409783   3.019391   
2015  35.576785   6.505772  3.765487  16.453165     4.957899   2.140953   
2016  40.550672   6.064707  4.280912  12.706759     5.609500   2.375809   
2017  37.649177   4.932482  3.694629  13.895294     5.306292   2.229850   
2018  35.697983   4.737251  3.897099  12.967288     5.808107   2.051611   
2019  23.094062   4.460207  3.888492  11.162451     6.548418   1.957615   
2020  14.680459   3.706338  2.436468   8.428774     4.455627   1.227336   

      Thailand  
2005  6.558305  
2006  6.184972  
2007  5.429818  
2008  4.589659  
2009  3.217705  
2010  4.090423  
2011  2.629800  
2012  3.499387  
2013  3.326040  
2014  3.501676  
2015  2.096523  
2016  3.853385  
2017  2.957005  
2018  2.542066  
2019  3.186669  
2020  1.458651

### Trial End